### PR TITLE
Upgrade `equatable_if_let` to style

### LINKED ISSUE #1716: Lint if lets that could be better worded as equality
This would also affect the `single_match` lint. Essentially, whenever someone writes a pattern like this:

```
if let Ordering::Greater = a.cmp(b) { ... }
```

It would be a bit clearer to write something like this:

```
if a.cmp(b) == Ordering::Greater { ... }
```

Specifically, if the left hand side of the `if let` contains a value that can be constructed directly, and the destructured value implements `PartialEq`, then is makes a lot more sense to refactor the `if let` into a direct comparison.

This would also affect the `single_match` and `single_match_else` lints to give proper suggestions, i.e. changing:

```
match a.cmp(b) {
    Ordering::Greater => { ... },
    _ => {},
}
```

to:

```
if a.cmp(b) == Ordering::Greater { ... }
```

### PR DESCRIPTION
**Pull request by HKalbasi**
Continue of #7762

Fix #1716

changelog: upgrade `equatable_if_let` to style and add pedantic lint `equatable_matches`

After reading some more about `StructuralPartialEq` I got that my previous understanding of it was completely wrong and it should be checked recursively. So now it can detect which pattern supports equality and that false positive is fixed.

Also I added a separate lint for `matches!` based on [this comment](https://github.com/rust-lang/rust-clippy/pull/7762#issuecomment-933707585). Lint for `if let` is style and lint for `matches!` is pedantic, but probably both of them need discussion.

### CODE REVIEW

**Comment by HKalbasi** [comment]
What about
```rust
if Some(42) == some_very.long().expr.chain[1..4].that_is().hard_to.read() { ... }
```
in this case? And also what do you think about
```rust
if let 42 = some_very.long().expr.chain[1..4].that_is().hard_to.read() { ... }
```
I'm agree that in the long case it doesn't make much difference. In these cases bottleneck of readablity is somewhere else.

Also there is a good amount of data in the dogfooding fixes of previous PR which we can see for deciding.

By the way, I don't now why CI is unhappy. It pass tests on my local system.

**Comment by flip1995** [comment]
> What about
> 
> ```rust
> if Some(42) == some_very.long().expr.chain[1..4].that_is().hard_to.read() { ... }
> ```
> 
> in this case? And also what do you think about
> 
> ```rust
> if let 42 = some_very.long().expr.chain[1..4].that_is().hard_to.read() { ... }
> ```

The first would go against the "yoda condition" argument in the lint documentation. It might be better, but the difference is 2 characters. Maybe it's just that I'm more used to seeing `if let Some(_) ...` and that's why this looks more natural to me. I just feel like this lint is pretty _pedantic_.

The second example looks weird to me, because you usually don't pattern match on just integers. 

Looking thorugh the dogfood fixes, I especially liked the changes from `if let Ordering::Greater = _` to `if _ == Ordering::Greater` and the like. I think only applying this lint on unit-variants and primitives can be warn-by-default. I don't really like it for more complex matches. pattern matching is such a core concept in Rust, that `if let E::ABC(4, 2) = expr` just feels more natural to me than `if E::ABC(4, 2) == expr`.

Matching on a unit-variant is more like comparing (`==`) values (like integers), but matching on more complex things is more like pattern matching to me (`if let`).

**Comment by Manishearth** [comment]
Going to r? @flip1995 on this one since he's already posted a bit. I agree with him on this though.

**Comment by camsteffen** [comment]
I like the idea of `style` for unit-variants.

I have a theory that this may be a minor perf issue, if only for unoptimized builds. Using `==` requires two instances of a type where pattern matching is just reading (the discriminant of) a single instance.

```rust
enum MyEnum {
    LittleGuy,
    BigGuy([u8; 100]),
}
```

`x == MyEnum::LittleGuy` may be less optimal than `if let MyEnum::LittleGuy = x` in unoptimized builds. I think this could be reason enough to only lint (by default) on enums with *all* unit-like variants. And also reason enough to make anything else `restriction`.

**Comment by flip1995** [comment]
I don't really see a blocker of a perf impact on unoptimized builds, as long as there is no perf impact for optimzed builds. So if the only argument for not putting the unit-variant-only version into style, then I would say put it into style anyway. That is, if we really want to split up the lint on the unit-variant boundary. If we want to keep everything in one lint, I don't have a strong feeling about pedantic vs restriction.

**Comment by HKalbasi** [comment]
> split up the lint on the unit-variant boundary

Or maybe a sensitivity configuration with some levels like "primitive", "unit-variant" and "all", with "unit-variant" as default?

**Comment by camsteffen** [comment]
> I don't really see a blocker of a perf impact on unoptimized builds

Well maybe my "unoptimized" point is kinda moot - that is actually a common tradeoff for abstractions in Rust.

> Also I added a separate lint for `matches!`

Just noticed this. Don't forget to write a "changelog:".

> Or maybe a sensitivity configuration with some levels like "primitive", "unit-variant" and "all", with "unit-variant" as default?

I guess that makes sense considering that it affects multiple lints? I'd be okay with that and `pedantic`.

**Comment by HKalbasi** [comment]
I added a configuration.

> Just noticed this. Don't forget to write a "changelog:".

There is a changelog, but it needs update after we decide about style and pedantic.

File: clippy_utils/src/higher.rs [hunk 1]
```diff
@@ -470,6 +470,42 @@ pub fn extract_assert_macro_args<'tcx>(e: &'tcx Expr<'tcx>) -> Option<Vec<&'tcx
...
+pub struct MatchesExpn<'tcx> {
+    /// Span of `matches!(..)`
+    pub call_site: Span,
+    /// Second parameter of `matches!`
+    pub arm: &'tcx Arm<'tcx>,
+    /// First parameter of `matches!`
+    pub exp: &'tcx Expr<'tcx>,
+}
+
+impl MatchesExpn<'tcx> {
```
Comment (flip1995) [inline clippy_utils/src/higher.rs:484]: Ah, we don't really want to keep doing this: #7843

Is this only required to generate the suggestion? If so, I'd prefer to just have a non-auto-applicable suggestion in the `matches` case with `Applicability::HasPlaceholder`

File: clippy_lints/src/utils/conf.rs [hunk 2]
```diff
@@ -296,6 +303,10 @@ define_Conf! {
     ///
     /// Whether to apply the raw pointer heuristic to determine if a type is `Send`.
     (enable_raw_pointer_heuristic_for_send: bool = true),
+    /// Lint: EQUATABLE_IF_LET, EQUATABLE_MATCHES
+    ///
+    /// Whether to detect pattern as a value for using with equality.
```
Comment (flip1995) [inline clippy_lints/src/utils/conf.rs:308]:
```suggestion
    /// Specifies which kind of patterns should be linted. Possible values are "All, "Simple", and "Primitive".
```

File: clippy_utils/src/ast_utils.rs [hunk 3]
```diff
@@ -278,7 +278,7 @@ pub fn eq_item_kind(l: &ItemKind, r: &ItemKind) -> bool {
         },
         (Trait(box TraitKind(la, lu, lg, lb, li)), Trait(box TraitKind(ra, ru, rg, rb, ri))) => {
             la == ra
-                && matches!(lu, Unsafe::No) == matches!(ru, Unsafe::No)
+                && matches!(lu, Unsafe::Yes(_)) == matches!(ru, Unsafe::Yes(_))
```
[the remark this answers is not in the document]
Comment (HKalbasi) [reply clippy_utils/src/ast_utils.rs:281]: I at first changed it to `lu == ru` which was wrong (I assumed there is `Unsafe::No` and `Unsafe::Yes`), and this style better shows why it is wrong so next person won't try `lu == ru`.

File: clippy_utils/src/higher.rs (hunk 1 above)
Comment (HKalbasi) [reply clippy_utils/src/higher.rs:484]: Parameters of `matches!` macro are needed to detect if this pattern is equatable.

**Comment by HKalbasi** [comment]
I made it pedantic, but style with sensitivity = primitive is also an option. I think it is clear that `if let "foo" = bar` is bad style.

**Comment by flip1995** [comment]
> I made it pedantic, but style with sensitivity = primitive is also an option. I think it is clear that `if let "foo" = bar` is bad style.

I agree. I think the `Simple` config value is the better default though. So for me the question is, if style+primitive or pedantic+simple is the better option.

File: clippy_utils/src/higher.rs (hunk 1 above)
Comment (flip1995) [reply clippy_utils/src/higher.rs:484]: Ah, that's unfortunate. We really have to figure out how to do this without relying on the exact expansion of the macro...

**Comment by camsteffen** [comment]
Yet another option is to make a `if_let_primitive` lint for `style`. I think I'd like that option best. My second pick would be simple+pedantic. The biggest downside IMO is with the primitive+style option because a lot of users won't discover the config or care to change it.

Also I would put "[aggregate primitives](https://github.com/rust-lang/rust-clippy/blob/84a4ab7f3f21ae78cd1c274c4647fd38eef7d43d/clippy_utils/src/ty.rs#L232-L237)" in the "simple" category. Don't know if that has been considered.

**Comment by HKalbasi** [comment]
I don't know how many people use pedantic lints, but primitive+style would be better for default users than simple+pedantic. `if_let_primitive` (and possibly `matches_primitive`) are duplicate lints to me, but they will solve the problem of noise/visibility.

> Also I would put "aggregate primitives" in the "simple" category. Don't know if that has been considered.

There are currently in "all".

**Comment by blyxyas** [comment]
Closing this because the original issue (#1716) was resolved by #9368

**Pull request closed without merging**